runtime/src/main/java/org/corfudb/runtime/collections/TxnContext.java

```diff
@@ -0,0 +1,517 @@
+        table.getMetrics().numScans.incrementAndGet();
+        return table.scanAndFilterByEntry(entryPredicate);
+    }
+
+    /**
+     * Start the actual corfu transaction if started by a read operation,
+     * Cases to consider:
+     *   Bit2      Bit1         Bit0
+     *  isRead   hasTxnBegun   isEmptyWriteBuffer
+     *  ======= ===========   ================
+     * 0 commit noTxnStarted  buffered writes => start txn, apply writes, writeOnlyTxn
+     * 1 commit noTxnStarted  no writes       => start txn, writeOnlyTxn
+     * 2 commit txnStarted    buffered writes => apply writes, writeAndReadTxn
+     * 3 commit txnStarted    no writes       => do nothing!
+     * 4 readApi noTxnStarted buffered writes => start txn, apply writes, writeAndReadTxn
+     * 5 readApi noTxnStarted no writes       => start txn, readOnlyTxn
+     * 6 readApi txnStarted   buffered writes => apply writes, writeAndReadTxn
+     * 7 readApi txnStarted   no writes       => readOnlyTxn iff not writeAndReadTxn
+     *
+     * @param tableBeingRead - am i being called from read api
+     * @param <K>
+     * @param <V>
+     * @param <M>
+     * @param tableBeingRead - only read transactions set this value
+     */
+    private <K extends Message, V extends Message, M extends Message>
+    void txBeginAndApply(Table<K, V, M> tableBeingRead) {
+        int decision = 0x0;
+        if (tableBeingRead != null) {
+            decision |= 0x04;
+        }
+        if (TransactionalContext.isInTransaction()) {
+            decision |= 0x02;
+        }
+        if (operations.isEmpty()) {
+            decision |= 0x01;
+        }
+        switch (decision) {
+            case 0: // commit() on a normal write-only transaction!
+                log.trace("commit() applying {} operations", operations.size());
+                txBeginInternal();
+                operations.forEach(Runnable::run);
+                operations.clear();
+                txnType = TxnType.writeOnly;
+                break;
+            case 1: // commit() called without a corfu transaction or buffered writes!
+                log.warn("commit called on an empty transaction!");
+                break;
+            case 2: // commit() called but corfu txn already begun for reads
+                log.trace("commit() applying pending {} operations after reads", operations.size());
+                operations.forEach(Runnable::run);
+                operations.clear();
+                txnType = TxnType.writeAndReads;
+                break;
+            case 3: // commit called in a read-only transaction.
+                if (txnType == TxnType.readOnly) {
+                    log.trace("Ending a read-only transaction");
+                } else if (txnType == TxnType.writeAndReads) {
+                    log.trace("Ending a dirty-read transaction!");
+                } else { // Attempting to commit a corfu txn started outside this scope!
```

Comment:
this means it is a write-only transaction and the transaction has already started.  How does it mean it is outside the scope?

Reply:
if commit() is called and no dirty reads have happened then that means corfu transaction should NOT have started.
But if we are in this case it means someone started a transaction out side of the TxnContext - that is not allowed and so we should throw an error.
That's why these cases are important - they highlight issues like this.